### ISSUE TITLE
change color of vehicle to visually indicate data propagation

### ISSUE THREAD

**Issue by myleskeller** [myleskeller/Carla-CPS#1]
https://github.com/myleskeller/Carla-CPS/blob/cb7fa78be2b4238fc23b74de2ed933e041c546c8/manual_control_AObstacleDetectionSensor.py#L1025

**Comment by myleskeller** [comment]
so I was able to get vehicles to respawn with their color changed to black (as an example), but I have had no success in figuring out how to get them to start moving again. I don't know what I'm doing wrong, but it probably has something to do with autopilot and the traffic manager.

below is the code snippet that accomplished this. it is inside `_on_obstacle()` in the `AObstacleDetectionSensor` class declaration.

```
                print("copying awaynet info...")
                # timestamp = AWayNet_list[other_actor].getTimestamp()
                # message = AWayNet_list[other_actor].getMessage()
                # del AWayNet_list[other_actor]

                print("redefining blueprint...")
                blueprint = random.choice(self.world.get_blueprint_library().filter(event.other_actor.type_id))
                if blueprint.has_attribute("color"):
                    color = "0,0,0"
                    blueprint.set_attribute("color", color)
                if blueprint.has_attribute("driver_id"): #? i have no idea what this is for...
                    driver_id = random.choice(blueprint.get_attribute("driver_id").recommended_values)
                    blueprint.set_attribute("driver_id", driver_id)
                    print("set 'driver_id'...")
                else:
                    blueprint.set_attribute("role_name", "autopilot")
                    print("set 'role_name' to 'autopilot'...")

                print("respawning black (hopefully) vehicle...")
                spawn_point = event.other_actor.get_transform()
                spawn_point.location.z += 5.0
                spawn_point.rotation.roll = 0.0
                spawn_point.rotation.pitch = 0.0

                event.other_actor.set_autopilot(False)
                event.other_actor.destroy()
                event.other_actor = self.world.try_spawn_actor(blueprint, spawn_point) #! ambiguous error idk...

                carla.command.SetAutopilot(event.other_actor, True, tm_port)
                event.other_actor.set_autopilot(True)
                self.show_vehicle_telemetry = False
                self.modify_vehicle_physics(event.other_actor)
                self.obstacle_sensor = AObstacleDetectionSensor(event.other_actor)

                print("re-adding awaynet info...")
                # AWayNet_list[other_actor] = AWayNet(message, timestamp)

                # print("ticking world...")
                # if self.world.sync:
                #     self.world.tick()
                # else:
                #     self.world.wait_for_tick()
```

**Comment by myleskeller** [comment]
For the time being, I'm settling for opening the doors and turning on all of the vehicle lights as sufficient indication of propagation until I can find another viable visual indicator or get this working. 

I also toyed around with attaching objects about 1 meter above each actor to indicate data propagation, but something about their positioning doesn't seem to use the same coordinate system as the vehicle actors and creates absolute chaos that eventually crashes the server. See below for how well using the traffic cone asset went:

https://user-images.githubusercontent.com/43857093/226479587-0afaa467-0c43-4d22-98e5-8e5a39882853.mp4

**Comment by myleskeller** [comment]
the culminating solution to visual indication of data propagation has been determined. and it's not changing vehicle colors.

`world.debug` has a bunch of useful geometry drawing instructions that can be used to create 3d shapes, lines, and render text in the environment in 3d. in the end, I used `debug.draw_point` to render cubes about 1 meter above each actor participating in the network. The color green was used to indicate "contacted" where red was for "not yet contacted".

https://user-images.githubusercontent.com/43857093/227744012-adf8f4ba-4692-4c17-a155-39ec4eb66f93.mp4

Furthermore, I ended up using `debug.draw_string` to show when (simulation timestamp) and where (string just lingers at contact coordinates) each propagation occurs.
